storage/remote/codec.go

```diff
@@ -502,6 +502,66 @@ func exemplarProtoToExemplar(ep prompb.Exemplar) exemplar.Exemplar {
 	}
 }
 
+func histogramProtoToHistogram(hp prompb.Histogram) histogram.Histogram {
+	h := histogram.Histogram{
+		Schema:          hp.Schema,
+		ZeroThreshold:   hp.ZeroThreshold,
+		ZeroCount:       hp.GetZeroCountInt(),
+		Count:           hp.GetCountInt(),
+		Sum:             hp.Sum,
+		PositiveBuckets: hp.PositiveBuckets.GetDelta(),
+		NegativeBuckets: hp.NegativeBuckets.GetDelta(),
+	}
+
+	// The following values can be nil and will cause a panic if accessed
+	// without being checked.
```

Comment:
My two cents here is that I slightly prefer how it is in `histogram.Histogram` because to me buckets and spans are pretty separate, so making them separate fields in the top level of the struct makes sense, and the names `..Spans` for an array of spans and `...Buckets` for an array of bucket deltas are clearer than in the proto, where it's "buckets but not an array containing bucket values, it's actually another struct with a field containing an array of spans which aren't buckets and then the actual bucket values but they're called deltas." Also `histogram.Histogram` is slightly easier to work with because there's no nested struct.

Reply:
_Disclaimer: This comment is not relevant to this PR but a different discussion._

I kind of like the layout we have in protobuf, precisely because the deltas between bucket counts are not really buckets. A bucket is a count with an upper and lower limit, and we encode that as a position in a series of spans and as deltas between the bucket counts. So arguably, the closest we get to a representation of buckets is indeed the entirety of the spans slice and the deltas slice. Another advantage I see with the layout we have right now in protobuf that we only have one field with a `Positive...` and one with a `Negative...` prefix (rather than two of each).

One could even argue (@bboreham did so) that we should push the bucket deltas down into the spans themselves, i.e. at the top level would be a span slice only, where each span contains an offset and a deltas slice of the length of the span. That's semantically really clear, but it's not the way we store it, and also a bit less efficient on the wire.

Which brings me to the efficiency argument. I think on the pure Go level, it doesn't really make a difference. Different packing order in RAM, but no different size. Maybe the nested structure is more effort to handle (for computers, but arguably also for the human programmer, as you said above). For protobuf, it's a bit harder to answer. I believe the flat layout would save us one tag.

I would very much like to bring things in line, so either use the nested layout or the flat layout both in protobuf as well as in Go.

While I tend to prefer the nested layout for the reasons stated above, I hear you that you like the flat layout more. Since the flat layout may have an efficiency gain for protobuf, let's first experiment with a flat layout for protobuf. I'll create a PR to discuss.

Reply:
Cf. https://github.com/prometheus/prometheus/pull/11011 (as you can see, PR numbers are now all binary…)